### PR TITLE
Support onChange callback in FluentUI Datepicker Input Field

### PR DESCRIPTION
**Pull request by rymeskar**
#### Pull request checklist

- [ ] Addresses an existing issue: Fixes #0000
- [ ] Include a change request file using `$ yarn change`

#### Description of changes

(give an overview)

#### Focus areas to test

(optional)

### CODE REVIEW

File: packages/fluentui/react-northstar/src/components/Datepicker/DatepickerCalendar.tsx [hunk 1]
```diff
@@ -309,20 +235,47 @@ DatepickerCalendar.propTypes = {
...
   accessibility: datepickerCalendarBehavior,
-  firstDayOfWeek: DayOfWeek.Monday,
+  firstDayOfWeek: DayOfWeek.Sunday,
   firstWeekOfYear: FirstWeekOfYear.FirstDay,
   dateRangeType: DateRangeType.Day,
   header: {},
   calendarCell: {},
   calendarHeaderCell: {},
-  localizedStrings: DEFAULT_CALENDAR_LOCALIZED_STRINGS,
+  ...DEFAULT_CALENDAR_STRINGS,
```
Comment (pompomon) [inline packages/fluentui/react-northstar/src/components/Datepicker/DatepickerCalendar.tsx:278]: Please could you update DatepickerCalendarHeader.tsx with the localized string for next/previous month (there are TODO: marks)

File: packages/fluentui/react-northstar/src/components/Datepicker/Datepicker.tsx [hunk 2]
```diff
@@ -234,17 +240,24 @@ export const Datepicker: ComponentWithAs<'div', DatepickerProps> &
             ...unhandledProps,
           })}
         >
-          <Input readOnly onClick={showCalendarGrid} value={valueFormatter(selectedDate)} />
+          <Input
```
Comment (pompomon) [inline packages/fluentui/react-northstar/src/components/Datepicker/Datepicker.tsx:243]: Please could you update this part to also use `createShorthand`, like other parts are doing?

File: packages/fluentui/react-northstar/src/components/Datepicker/Datepicker.tsx [hunk 3]
```diff
@@ -165,31 +94,65 @@ export const Datepicker: ComponentWithAs<'div', DatepickerProps> &
   const context = useFluentContext();
   const { setStart, setEnd } = useTelemetry(Datepicker.displayName, context.telemetry);
   setStart();
+  const dateSelectionErrorString = 'A date selection is required';
   const datepickerRef = React.useRef<HTMLElement>();
-  const [open, setOpen] = React.useState<boolean>(false);
+  const [openState, setOpenState] = useAutoControlled<OpenState>({
```
Comment (layershifter) [inline packages/fluentui/react-northstar/src/components/Datepicker/Datepicker.tsx:99]: This should be just `boolean`

File: packages/fluentui/react-northstar/src/components/Datepicker/Datepicker.tsx [hunk 4]
```diff
@@ -165,31 +94,65 @@ export const Datepicker: ComponentWithAs<'div', DatepickerProps> &
...
+    defaultValue: props.defaultCalendarOpenState ? OpenState.Open : OpenState.Closed,
+    value: undefined,
+    initialValue: OpenState.Closed,
```
Comment (layershifter) [inline packages/fluentui/react-northstar/src/components/Datepicker/Datepicker.tsx:102]:
```suggestion
    initialValue: false,
```

File: packages/fluentui/react-northstar/src/components/Datepicker/Datepicker.tsx [hunk 5]
```diff
@@ -165,31 +94,65 @@ export const Datepicker: ComponentWithAs<'div', DatepickerProps> &
   const context = useFluentContext();
   const { setStart, setEnd } = useTelemetry(Datepicker.displayName, context.telemetry);
   setStart();
+  const dateSelectionErrorString = 'A date selection is required';
   const datepickerRef = React.useRef<HTMLElement>();
-  const [open, setOpen] = React.useState<boolean>(false);
+  const [openState, setOpenState] = useAutoControlled<OpenState>({
+    defaultValue: props.defaultCalendarOpenState ? OpenState.Open : OpenState.Closed,
```
Comment (layershifter) [inline packages/fluentui/react-northstar/src/components/Datepicker/Datepicker.tsx:100]:
```suggestion
    defaultValue: props.defaultOpen,
```

File: packages/fluentui/react-northstar/src/components/Datepicker/Datepicker.tsx [hunk 6]
```diff
@@ -165,31 +94,65 @@ export const Datepicker: ComponentWithAs<'div', DatepickerProps> &
...
+    value: undefined,
```
Comment (layershifter) [inline packages/fluentui/react-northstar/src/components/Datepicker/Datepicker.tsx:101]:
```suggestion
    value: props.open,
```

File: packages/fluentui/react-northstar/src/components/Datepicker/Datepicker.tsx [hunk 7]
```diff
@@ -165,31 +94,65 @@ export const Datepicker: ComponentWithAs<'div', DatepickerProps> &
...
-  const calendarOptions: IDayGridOptions = {
-    selectedDate: selectedDate ?? props.today ?? new Date(),
-    navigatedDate: selectedDate ?? props.today ?? new Date(),
-    firstDayOfWeek,
-    firstWeekOfYear,
-    dateRangeType,
-  };
+  const [formattedDate, setFormattedDate] = React.useState<string>('');
+  const [error, setError] = React.useState<string>(() =>
+    props.isRequired && !selectedDate ? dateSelectionErrorString : '',
```
Comment (layershifter) [inline packages/fluentui/react-northstar/src/components/Datepicker/Datepicker.tsx:107]:
```suggestion
    props.required && !selectedDate ? dateSelectionErrorString : '',
```

We don't use `is` prefix

File: packages/fluentui/react-northstar/src/components/Datepicker/Datepicker.tsx [hunk 8]
```diff
@@ -165,31 +94,65 @@ export const Datepicker: ComponentWithAs<'div', DatepickerProps> &
...
+  );
 
-  const showCalendarGrid = () => {
-    setOpen(true);
+  const { calendar, popup, className, design, styles, variables, formatMonthDayYear } = props;
+  const valueFormatter = date => (date ? formatMonthDayYear(date) : '');
```
Comment (layershifter) [inline packages/fluentui/react-northstar/src/components/Datepicker/Datepicker.tsx:111]: Please move out it from component's body

File: packages/fluentui/react-northstar/src/components/Datepicker/Datepicker.tsx [hunk 9]
```diff
@@ -165,31 +94,65 @@ export const Datepicker: ComponentWithAs<'div', DatepickerProps> &
...
+    daysToSelectInDayView: props.daysToSelectInDayView,
+    today: props.today,
+    showWeekNumbers: props.showWeekNumbers,
+    workWeekDays: props.workWeekDays,
+    minDate: props.minDate,
+    maxDate: props.maxDate,
+    restrictedDates: props.restrictedDates,
+  } as IDayGridOptions;
+
+  const dateFormatting = {
```
Comment (layershifter) [inline packages/fluentui/react-northstar/src/components/Datepicker/Datepicker.tsx:130]: Is there a type for this?

File: packages/fluentui/react-northstar/src/components/Datepicker/Datepicker.tsx [hunk 10]
```diff
@@ -213,10 +176,15 @@ export const Datepicker: ComponentWithAs<'div', DatepickerProps> &
 
   const overrideDatepickerCalendarProps = (predefinedProps: DatepickerCalendarProps): DatepickerCalendarProps => ({
     ...calendarOptions,
+    ...dateFormatting,
     onDateChange: (e, itemProps) => {
-      setSelectedDate(itemProps.value.originalDate);
-      setOpen(false);
-      _.invoke(predefinedProps, 'onDateChange', e, itemProps);
+      const targetDay = itemProps.value;
+      setSelectedDate(targetDay.originalDate);
```
Comment (layershifter) [inline packages/fluentui/react-northstar/src/components/Datepicker/Datepicker.tsx:182]: Please switch state management to `useReducer`, it's hard to follow this

File: packages/fluentui/react-northstar/src/components/Datepicker/Datepicker.tsx [hunk 11]
```diff
@@ -165,31 +94,65 @@ export const Datepicker: ComponentWithAs<'div', DatepickerProps> &
...
+  );
 
-  const showCalendarGrid = () => {
-    setOpen(true);
+  const { calendar, popup, className, design, styles, variables, formatMonthDayYear } = props;
+  const valueFormatter = date => (date ? formatMonthDayYear(date) : '');
+
+  const nonNullSelectedDate = selectedDate ?? props.today ?? new Date();
+
+  const calendarOptions = {
```
Comment (layershifter) [inline packages/fluentui/react-northstar/src/components/Datepicker/Datepicker.tsx:115]: Is there a type for this?

File: packages/fluentui/react-northstar/src/components/Datepicker/Datepicker.tsx [hunk 12]
```diff
@@ -225,6 +193,44 @@ export const Datepicker: ComponentWithAs<'div', DatepickerProps> &
...
+      } else {
+        setOpenState(OpenState.Open);
+      }
+    }
+  };
+
+  const onInputChange = (e, target: { value: string }) => {
+    const parsedDate = props.parseDate(target.value);
+
+    setFormattedDate(target.value);
```
Comment (layershifter) [inline packages/fluentui/react-northstar/src/components/Datepicker/Datepicker.tsx:216]: Validation is pure and should be done outside of a component

File: packages/fluentui/react-northstar/src/components/Datepicker/Datepicker.tsx [hunk 13]
```diff
@@ -276,25 +296,49 @@ Datepicker.propTypes = {
...
+  shortMonths: PropTypes.arrayOf(PropTypes.string),
+  days: PropTypes.arrayOf(PropTypes.string),
+  shortDays: PropTypes.arrayOf(PropTypes.string),
+
+  isRequiredErrorMessage: PropTypes.string,
+  invalidInputErrorMessage: PropTypes.string,
+  isOutOfBoundsErrorMessage: PropTypes.string,
   goToToday: PropTypes.string,
-  placeholder: PropTypes.string,
+  prevMonthAriaLabel: PropTypes.string,
```
Comment (layershifter) [inline packages/fluentui/react-northstar/src/components/Datepicker/Datepicker.tsx:315]: Why all of these are not in `LocalizedStrings`?

File: packages/fluentui/react-northstar/src/components/Datepicker/Datepicker.tsx [hunk 14]
```diff
@@ -264,6 +277,13 @@ Datepicker.propTypes = {
   calendar: customPropTypes.itemShorthand,
   popup: customPropTypes.itemShorthand,
 
+  disabled: PropTypes.bool,
+  isRequired: PropTypes.bool,
+  onDateChange: PropTypes.func,
+  placeholder: PropTypes.string,
+  allowTextInput: PropTypes.bool,
```
Comment (layershifter) [inline packages/fluentui/react-northstar/src/components/Datepicker/Datepicker.tsx:284]: @pompomon is there a better name for this prop?

File: packages/fluentui/react-northstar/src/components/Datepicker/Datepicker.tsx (hunk 2 above)
Comment (rymeskar) [reply packages/fluentui/react-northstar/src/components/Datepicker/Datepicker.tsx:243]: Should I then also add slots for Input?

File: packages/fluentui/react-northstar/src/components/Datepicker/Datepicker.tsx (hunk 13 above)
Comment (rymeskar) [reply packages/fluentui/react-northstar/src/components/Datepicker/Datepicker.tsx:315]: @layershifter Do I understand that I should rather add the localizedStrings type as property and not extend the localizedStrings type?

File: packages/fluentui/react-northstar/src/components/Datepicker/Datepicker.tsx (hunk 14 above)
Comment (rymeskar) [reply packages/fluentui/react-northstar/src/components/Datepicker/Datepicker.tsx:284]: @layershifter  I copied the name from the [FabricUI](https://github.com/microsoft/fluentui/blob/fdc095fd93f528381650e1215bcf2bfe1ff9acf6/packages/date-time/src/components/DatePicker/DatePicker.base.tsx#L477).

File: packages/fluentui/react-northstar/src/components/Datepicker/Datepicker.tsx (hunk 12 above)
Comment (rymeskar) [reply packages/fluentui/react-northstar/src/components/Datepicker/Datepicker.tsx:216]: @layershifter just outside of the component's body or completely outside of this file?

File: packages/fluentui/react-northstar/src/components/Datepicker/Datepicker.tsx (hunk 11 above)
Comment (rymeskar) [reply packages/fluentui/react-northstar/src/components/Datepicker/Datepicker.tsx:115]: Added type.

File: packages/fluentui/react-northstar/src/components/Datepicker/Datepicker.tsx (hunk 9 above)
Comment (rymeskar) [reply packages/fluentui/react-northstar/src/components/Datepicker/Datepicker.tsx:130]: Added type.

File: packages/fluentui/react-northstar/src/components/Datepicker/Datepicker.tsx (hunk 8 above)
Comment (rymeskar) [reply packages/fluentui/react-northstar/src/components/Datepicker/Datepicker.tsx:111]: @layershifter  in this same file or in different one?

File: packages/fluentui/react-northstar/src/components/Datepicker/Datepicker.tsx (hunk 7 above)
Comment (rymeskar) [reply packages/fluentui/react-northstar/src/components/Datepicker/Datepicker.tsx]: Fixed.

File: packages/fluentui/react-northstar/src/components/Datepicker/Datepicker.tsx (hunk 6 above)
Comment (rymeskar) [reply packages/fluentui/react-northstar/src/components/Datepicker/Datepicker.tsx:101]: @layershifter  what would then be the difference between 'props.defaultOpen' and 'props.open'?

File: packages/fluentui/react-northstar/src/components/Datepicker/Datepicker.tsx (hunk 3 above)
Comment (rymeskar) [reply packages/fluentui/react-northstar/src/components/Datepicker/Datepicker.tsx:99]: @layershifter I use the enum to mimic the behavior of FabricUI/Teams. https://github.com/microsoft/fluentui/blob/0b666d9bc2e4b20e9267a78de4327998eda5f6b4/packages/fluentui/react-northstar/src/components/Datepicker/Datepicker.tsx#L204

Should I then simplify to boolean but lose the described behavior?

FYI: @pompomon

File: packages/fluentui/react-northstar/src/components/Datepicker/Datepicker.tsx (hunk 2 above)
Comment (pompomon) [reply packages/fluentui/react-northstar/src/components/Datepicker/Datepicker.tsx:243]: Yeah, let's do that :)

File: packages/fluentui/react-northstar/src/components/Datepicker/Datepicker.tsx (hunk 14 above)
Comment (pompomon) [reply packages/fluentui/react-northstar/src/components/Datepicker/Datepicker.tsx:284]: ManualInput/AllowManualInput was my other option, but I think we can do that like it

File: packages/fluentui/react-northstar/src/components/Datepicker/Datepicker.tsx [hunk 15]
```diff
@@ -225,6 +193,44 @@ export const Datepicker: ComponentWithAs<'div', DatepickerProps> &
...
+    }
+  };
+
+  const onInputChange = (e, target: { value: string }) => {
+    const parsedDate = props.parseDate(target.value);
+
+    setFormattedDate(target.value);
+    if (parsedDate) {
+      if (isRestrictedDate(parsedDate, calendarOptions)) {
+        setError('The selected date is from the restricted range.');
```
Comment (pompomon) [inline packages/fluentui/react-northstar/src/components/Datepicker/Datepicker.tsx:219]: That should be a prop, so consumers can put the localized version of the error

File: packages/fluentui/react-northstar/src/components/Datepicker/Datepicker.tsx [hunk 16]
```diff
@@ -225,6 +193,44 @@ export const Datepicker: ComponentWithAs<'div', DatepickerProps> &
...
+      } else {
+        setError('');
+        setSelectedDate(parsedDate);
+        _.invoke(props, 'onDateChange', e, { ...props, value: parsedDate });
+      }
+    } else if (target.value) {
+      setError('Manually entered date is not in correct format.');
```
Comment (pompomon) [inline packages/fluentui/react-northstar/src/components/Datepicker/Datepicker.tsx:226]: Same here, we should be able to localize all the labels

Comment (rymeskar) [reply packages/fluentui/react-northstar/src/components/Datepicker/Datepicker.tsx:226]: Good point, changed.

File: packages/fluentui/react-northstar/src/components/Datepicker/Datepicker.tsx (hunk 15 above)
Comment (rymeskar) [reply packages/fluentui/react-northstar/src/components/Datepicker/Datepicker.tsx:219]: Good point, changed.

File: packages/fluentui/react-northstar/src/components/Datepicker/Datepicker.tsx (hunk 10 above)
Comment (rymeskar) [reply packages/fluentui/react-northstar/src/components/Datepicker/Datepicker.tsx:185]: Is it then fair to also create a composite state object { error, formattedDate, selectedDate }  or you would keep the states separate?

File: packages/fluentui/docs/src/components/ComponentPlayground/componentGenerators.tsx [hunk 17]
```diff
@@ -36,6 +37,9 @@ export const Box: KnobComponentGenerators<BoxProps> = {
   children: () => null,
 };
 
+export const Input: KnobComponentGenerators<InputProps> = {
+  children: () => null,
+};
```
Comment (layershifter) [inline packages/fluentui/docs/src/components/ComponentPlayground/componentGenerators.tsx:42]: It should not be required.

File: packages/fluentui/react-northstar/src/components/Datepicker/Datepicker.tsx (hunk 3 above)
Comment (layershifter) [reply packages/fluentui/react-northstar/src/components/Datepicker/Datepicker.tsx:102]: I don't see any reason to have enum state, `open` is am `boolean` value

File: packages/fluentui/react-northstar/src/components/Datepicker/Datepicker.tsx (hunk 8 above)
Comment (layershifter) [reply packages/fluentui/react-northstar/src/components/Datepicker/Datepicker.tsx:114]: The same

File: packages/fluentui/react-northstar/src/components/Datepicker/Datepicker.tsx (hunk 10 above)
Comment (layershifter) [reply packages/fluentui/react-northstar/src/components/Datepicker/Datepicker.tsx:185]: > Is it then fair to also create a composite state object { error, formattedDate, selectedDate } or you would keep the states separate?

Let's keep as it is for now.

File: packages/fluentui/react-northstar/src/components/Datepicker/Datepicker.tsx (hunk 12 above)
Comment (layershifter) [reply packages/fluentui/react-northstar/src/components/Datepicker/Datepicker.tsx:219]: Probably even in a separate file. I expect something like this:

```
const { error } = validateDate()

setError(error)
```

File: packages/fluentui/react-northstar/src/components/Datepicker/Datepicker.tsx (hunk 3 above)
Comment (rymeskar) [reply packages/fluentui/react-northstar/src/components/Datepicker/Datepicker.tsx:102]: @layershifter switched in [the following commit](https://github.com/microsoft/fluentui/pull/14563/commits/f4c038eeed968dfec4a00cc20d2194bc72072387).

File: packages/fluentui/docs/src/components/ComponentPlayground/componentGenerators.tsx (hunk 17 above)
Comment (rymeskar) [reply packages/fluentui/docs/src/components/ComponentPlayground/componentGenerators.tsx:42]: When removed, I got the following error on viewing examples:
image

File: packages/fluentui/react-northstar/src/components/Datepicker/Datepicker.tsx (hunk 8 above)
Comment (rymeskar) [reply packages/fluentui/react-northstar/src/components/Datepicker/Datepicker.tsx:114]: Should I move the `valueFormatter` away from component's body even though it depends on props?

File: packages/fluentui/react-northstar/src/components/Datepicker/Datepicker.tsx (hunk 12 above)
Comment (rymeskar) [reply packages/fluentui/react-northstar/src/components/Datepicker/Datepicker.tsx:219]: Updated in [the following commit](https://github.com/microsoft/fluentui/pull/14563/commits/f1dce86c84c5559c3e397d00f4b20bd814c3fb7f).

File: packages/fluentui/react-northstar/src/components/Datepicker/Datepicker.tsx (hunk 8 above)
Comment (layershifter) [reply packages/fluentui/react-northstar/src/components/Datepicker/Datepicker.tsx:114]: I see, now it depends on props. Ignore it

File: packages/fluentui/react-northstar/src/components/Datepicker/Datepicker.tsx (hunk 13 above)
Comment (layershifter) [reply packages/fluentui/react-northstar/src/components/Datepicker/Datepicker.tsx:324]: Oh, I see. All strings are props actually... Let's discuss this offline
